From a5c2cb8dd3bf162565c0fbf412b41d1b6b1c170a Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Fri, 29 Aug 2014 18:39:57 -0400
Subject: [net] macvlan: Allow setting multicast filter on all macvlan types

Message-id: <1409337597-30071-1-git-send-email-vyasevic@redhat.com>
Patchwork-id: 89896
O-Subject: [RHEL7.1 PATCH] macvlan: Allow setting multicast filter on all macvlan types
Bugzilla: 848197
RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Amos Kong <akong@redhat.com>
RH-Acked-by: Jason Wang <jasowang@redhat.com>

From: Vlad Yasevich <vyasevic@redhat.com>

Bugzilla: 848197
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7886289

Upstream commit 8a50f11c3b176d7a1df8cd5e29cbe965905e51ee
Author: Vlad Yasevich <vyasevic@redhat.com>
Date:   Fri Aug 15 13:04:59 2014 -0400

    Currently, macvlan code restricts multicast and unicast
    filter setting only to passthru devices.  As a result,
    if a guest using macvtap wants to receive multicast
    traffic, it has to set IFF_ALLMULTI or IFF_PROMISC.

    This patch makes it possible to use the fdb interface
    to add multicast addresses to the filter thus allowing
    a guest to receive only targeted multicast traffic.

    CC: John Fastabend <john.r.fastabend@intel.com>
    CC: Michael S. Tsirkin <mst@redhat.com>
    CC: Jason Wang <jasowang@redhat.com>
    Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
    Acked-by: John Fastabend <john.r.fastabend@intel.com>
    Acked-by: Michael S. Tsirkin <mst@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/macvlan.c b/drivers/net/macvlan.c
index b71dda3..bddccbb 100644
--- a/drivers/net/macvlan.c
+++ b/drivers/net/macvlan.c
@@ -595,7 +595,10 @@ static int macvlan_fdb_add(struct ndmsg *ndm, struct nlattr *tb[],
  struct macvlan_dev *vlan = netdev_priv(dev);
  int err = -EINVAL;
 
- if (!vlan->port->passthru)
+ /* Support unicast filter only on passthru devices.
+  * Multicast filter should be allowed on all devices.
+  */
+ if (!vlan->port->passthru && is_unicast_ether_addr(addr))
   return -EOPNOTSUPP;
 
  if (is_unicast_ether_addr(addr))
@@ -613,7 +616,10 @@ static int macvlan_fdb_del(struct ndmsg *ndm, struct nlattr *tb[],
  struct macvlan_dev *vlan = netdev_priv(dev);
  int err = -EINVAL;
 
- if (!vlan->port->passthru)
+ /* Support unicast filter only on passthru devices.
+  * Multicast filter should be allowed on all devices.
+  */
+ if (!vlan->port->passthru && is_unicast_ether_addr(addr))
   return -EOPNOTSUPP;
 
  if (is_unicast_ether_addr(addr))
-- 
1.7.1